crates/hir-expand/src/builtin/attr_macro.rs

```diff
@@ -76,6 +79,27 @@ fn dummy_attr_expand(
     ExpandResult::ok(tt.clone())
 }
 
+fn dummy_gate_test_expand(
+    db: &dyn ExpandDatabase,
+    id: MacroCallId,
+    tt: &tt::Subtree,
+    span: Span,
+) -> ExpandResult<tt::Subtree> {
+    let loc = db.lookup_intern_macro_call(id);
+    let crate_graph = db.crate_graph();
+    let test_is_active =
+        crate_graph[loc.krate].cfg_options.check_atom(&CfgAtom::Flag(sym::test.clone()));
+    if test_is_active {
+        ExpandResult::ok(tt.clone())
+    } else {
+        let result = quote::quote! { span=>
+            #[cfg(test)]
+            #tt
```

Comment:
I specifically not returned empty token tree because this will mean users will lose the few IDE features that still work in cfg'ed out code: namely, syntax errors and diagnostic about code being cfg'ed out.

Reply:
We can choose not to, but I wanted to be extra safe.

Reply:
Actually we  can just unconditionally put `#[cfg(test)]` into the expansion